haskell-src/Concordium/MerkleProofs.hs

```diff
@@ -0,0 +1,566 @@
+                            Sequence
+                                [ Tagged "round" u64,
+                                  Tagged "finalizers" (VariableLengthBytes SizeU32BE)
+                                ],
+                          RepeatedBE "finalizerQCRoundsSecondEpoch" SizeU32BE $
+                            Sequence
+                                [ Tagged "round" u64,
+                                  Tagged "finalizers" (VariableLengthBytes SizeU32BE)
+                                ],
+                          Tagged "aggregateSignature" (FixedLengthBytes 48)
+                        ]
+                    )
+        epochFinalizationEntryHash <-
+            setFresh $
+                Choice
+                    ( Sequence
+                        [ LiteralBytes (S.encode (0 :: Word8)),
+                          Tagged "null" (FixedLengthBytes 0)
+                        ]
+                    )
+                    ( Sequence
+                        [ LiteralBytes (S.encode (1 :: Word8)),
+                          Hashed "finalizedBlock" blockHash,
+                          Tagged "finalizedRound" u64,
+                          Tagged "epoch" u64,
+                          Tagged "finalizedAggregateSignature" (FixedLengthBytes 48),
+                          Tagged "finalizedSignatories" (VariableLengthBytes SizeU32BE),
+                          Tagged "successorAggregateSignature" (FixedLengthBytes 48),
+                          Tagged "successorSignatories" (VariableLengthBytes SizeU32BE),
+                          Tagged "successorProof" opaqueHash
+                        ]
+                    )
+        timeoutFinalizationHash <-
+            setFresh . Sequence $
+                [ Hashed "timeoutCertificate" timeoutCertificateHash,
+                  Hashed "epochFinalizationEntry" epochFinalizationEntryHash
+                ]
+        certificatesHash <-
+            setFresh . node $
+                [ ("quorumCertificate", quorumCertificateHash),
+                  ("timeoutFinalization", timeoutFinalizationHash)
+                ]
+        metaHash <- setFresh . node $ [("bakerInfo", bakerInfoHash), ("certificatesHash", certificatesHash)]
+        transactionsAndOutcomesHash <-
+            setFresh . Sequence $
+                [ Tagged "transactions" opaqueHash,
+                  Tagged "outcomes" opaqueHash
+                ]
+        dataHash <-
+            setFresh . Sequence $
+                [ Hashed "transactionsAndOutcomes" transactionsAndOutcomesHash,
+                  Tagged "state" opaqueHash
+                ]
+        blockQuasiHash <- setFresh . Sequence $ [Hashed "meta" metaHash, Hashed "data" dataHash]
+        schemaAt blockHash ?= Sequence [Hashed "header" blockHeaderHash, Hashed "quasi" blockQuasiHash]
+        return blockHash
+
+-- | A class for types that can produce 'MerkleProof's.
+class MerkleProvable m t where
```

Comment:
Again, please document what the type parameters are meant to be, ideally naming them more meaningfully.